crates/rethnet_evm/src/mempool.rs

```diff
@@ -0,0 +1,161 @@
+        self.add_transaction_impl(state, transaction)
+    }
+
+    /// Removes the transaction corresponding to the provided transaction hash, if it exists.
+    pub fn remove_transaction(&mut self, hash: &B256) -> Option<PendingTransaction> {
+        if let Some((idx, _)) = self
+            .pending_transactions
+            .iter()
+            .enumerate()
+            .find(|(_, transaction)| *transaction.hash() == *hash)
+        {
+            return Some(self.pending_transactions.remove(idx));
+        }
+
+        if let Some((idx, _)) = self
+            .future_transactions
+            .iter()
+            .enumerate()
+            .find(|(_, transaction)| *transaction.hash() == *hash)
+        {
+            return Some(self.future_transactions.remove(idx));
+        }
+
+        None
+    }
+
+    /// Updates the [`Pool`], moving any future transactions to the pending status, if their nonces are high enough.
+    pub fn update<S>(&mut self, state: &S)
+    where
+        S: StateRef + ?Sized,
+        S::Error: Debug,
+    {
+        let mut future_transactions = Vec::with_capacity(self.future_transactions.capacity());
+        std::mem::swap(&mut self.future_transactions, &mut future_transactions);
+
+        for transaction in future_transactions.into_iter() {
+            self.add_transaction_impl(state, transaction)
+                .expect("All future transactions have already been validated");
+        }
+    }
+
+    /// Returns all pending transactions, for which the nonces are too high.
+    pub fn future_transactions(&self) -> &[PendingTransaction] {
+        &self.future_transactions
+    }
+
+    /// Returns all pending transactions, for which the nonces are guaranteed to be high enough.
+    pub fn pending_transactions(&self) -> &[PendingTransaction] {
+        &self.pending_transactions
+    }
+
+    /// Returns the pending transaction corresponding to the provided hash, if it exists.
+    pub fn transaction_by_hash(&self, hash: &B256) -> Option<&PendingTransaction> {
+        self.pending_transactions
+            .iter()
+            .find(|transaction| *transaction.hash() == *hash)
+            .or_else(|| {
+                self.future_transactions
+                    .iter()
+                    .find(|transaction| *transaction.hash() == *hash)
+            })
+    }
+
+    fn add_transaction_impl<S: StateRef + ?Sized>(
+        &mut self,
+        state: &S,
+        transaction: PendingTransaction,
+    ) -> Result<(), MinerTransactionError<S::Error>> {
+        let transaction_gas_limit = U256::from(transaction.gas_limit());
+        if transaction_gas_limit > self.block_gas_limit {
+            return Err(MinerTransactionError::ExceedsBlockGasLimit {
+                block_gas_limit: self.block_gas_limit,
+                transaction_gas_limit,
+            });
+        }
+
+        let account = state.basic(*transaction.caller())?;
+
+        // Question: Must the account exist?
```

Comment:
@fvictorio & @alcuadrado

Reply:
Every account exists, even if it hasn't been touched and has no state associated to it.